### PR TITLE
should close http.Body before retry

### PR DESCRIPTION
**Pull request by jiuker**
**Please provide a description of this PR:**

should close http.Body before retry

### CODE REVIEW

File: pkg/spiffe/spiffe.go [hunk 1]
```diff
@@ -241,6 +241,7 @@ func RetrieveSpiffeBundleRootCerts(config map[string]string, caCertPool *x509.Ce
 			} else if resp.StatusCode != http.StatusOK {
 				b := make([]byte, 1024)
 				n, _ := resp.Body.Read(b)
+				_ = resp.Body.Close()
```
Comment (howardjohn) [inline pkg/spiffe/spiffe.go:244]: are we cllsing twice now? L260

**Review by hzxuzhonghu — COMMENTED** [review]
I am thinking we should use `NewExponentialBackOff`, then could use defer

**Comment by jiuker** [comment]
> NewExponentialBackOff

There seems to be a difference in effect with your recommended function, the original function will retry when the status is not Ok, but it also doesn't close the resp. Body at that point, and this PR is to fix that thing. But your recommended function will get something until it succeeds, which is different from getting something that succeeds with an exception.

**Comment by jiuker** [comment]
> I am thinking we should use `NewExponentialBackOff`, then could use defer



> I am thinking we should use `NewExponentialBackOff`, then could use defer

In fact, I don't see why this function could use defer.

**Comment by jiuker** [comment]
@howardjohn What are the suggestions for this one change. Body here does have the problem of not closing the resp. You will need to analyse this in depth.

**Review by hzxuzhonghu — APPROVED** [review]

File: pkg/spiffe/spiffe.go (hunk 1 above)
Comment (howardjohn) [reply pkg/spiffe/spiffe.go:244]: This still applies, I am not sure why this was resolved?

Comment (ymesika) [reply pkg/spiffe/spiffe.go:244]: @jiuker the Body closed is handled by the defer in line 259.
Why do you think we should close it in this block as well?
Please close the PR if you think it's redundant.

**Comment by jiuker** [comment]
No one seems to think it's a bug？

**Comment by jiuker** [comment]
> > No one seems to think it's a bug？
> 
> You never responded to the review comments: [#42581 (comment)](https://github.com/istio/istio/pull/42581#discussion_r1057783395)

```go
		var resp *http.Response
		for {
			resp, err = httpClient.Get(endpoint)
			var errMsg string
			if err != nil {
				errMsg = fmt.Sprintf("Calling %s failed with error: %v", endpoint, err)
			} else if resp == nil {
				errMsg = fmt.Sprintf("Calling %s failed with nil response", endpoint)
			} else if resp.StatusCode != http.StatusOK {
				b := make([]byte, 1024)
				n, _ := resp.Body.Read(b)
				errMsg = fmt.Sprintf("Calling %s failed with unexpected status: %v, fetching bundle: %s",
					endpoint, resp.StatusCode, string(b[:n]))
			} else {
				break
			}

			if startTime.Add(retryTimeout).Before(time.Now()) {
				return nil, fmt.Errorf("exhausted retries to fetch the SPIFFE bundle %s from url %s. Latest error: %v",
					trustDomain, endpoint, errMsg)
			}

			spiffeLog.Warnf("%s, retry in %v", errMsg, retryBackoffTime)
			time.Sleep(retryBackoffTime)
			retryBackoffTime *= 2 // Exponentially increase the retry backoff time.
		}
		defer resp.Body.Close()
```
What happens if we assume it returns something other than OK?
```diff
+		var resp *http.Response
		for {
+			resp, err = httpClient.Get(endpoint)
+			var errMsg string
			if err != nil {
				errMsg = fmt.Sprintf("Calling %s failed with error: %v", endpoint, err)
			} else if resp == nil {
				errMsg = fmt.Sprintf("Calling %s failed with nil response", endpoint)
			} else if resp.StatusCode != http.StatusOK {
+				b := make([]byte, 1024)
+				n, _ := resp.Body.Read(b)
+				errMsg = fmt.Sprintf("Calling %s failed with unexpected status: %v, fetching bundle: %s",
					endpoint, resp.StatusCode, string(b[:n]))
			} else {
				break
			}

+			if startTime.Add(retryTimeout).Before(time.Now()) {
				return nil, fmt.Errorf("exhausted retries to fetch the SPIFFE bundle %s from url %s. Latest error: %v",
					trustDomain, endpoint, errMsg)
			}

+			spiffeLog.Warnf("%s, retry in %v", errMsg, retryBackoffTime)
+			time.Sleep(retryBackoffTime)
+			retryBackoffTime *= 2 // Exponentially increase the retry backoff time.
		}
		defer resp.Body.Close()
```
Trying again now
```diff
		var resp *http.Response
		for {
+			resp, err = httpClient.Get(endpoint)
+			var errMsg string
			if err != nil {
				errMsg = fmt.Sprintf("Calling %s failed with error: %v", endpoint, err)
			} else if resp == nil {
				errMsg = fmt.Sprintf("Calling %s failed with nil response", endpoint)
			} else if resp.StatusCode != http.StatusOK {
				b := make([]byte, 1024)
				n, _ := resp.Body.Read(b)
				errMsg = fmt.Sprintf("Calling %s failed with unexpected status: %v, fetching bundle: %s",
					endpoint, resp.StatusCode, string(b[:n]))
			} else {
+				break
			}

			if startTime.Add(retryTimeout).Before(time.Now()) {
				return nil, fmt.Errorf("exhausted retries to fetch the SPIFFE bundle %s from url %s. Latest error: %v",
					trustDomain, endpoint, errMsg)
			}

			spiffeLog.Warnf("%s, retry in %v", errMsg, retryBackoffTime)
			time.Sleep(retryBackoffTime)
			retryBackoffTime *= 2 // Exponentially increase the retry backoff time.
		}
+		defer resp.Body.Close()
```
Doesn't it look like the first resp returned that wasn't OK was overwritten by the new resp returned? But its Body is required to be closed. This causes the resp before the retry to leak and not be closed. @howardjohn

**Pull request closed without merging**